src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -356,52 +356,56 @@ protected void CheckDocumentIsOpen(DocumentId documentId)
-                else
+                        // keep open document text alive by using PreserveIdentity
+                        return oldSolution.WithDocumentText(documentId, newTextAndVersion, PreservationMode.PreserveIdentity);
+                    }
+                    else
+                    {
+                        // We don't have the old text or version.  Rather than trying to reuse a version that we still have, let's just assume the file has changed.
+                        // keep open document text alive by using PreserveIdentity
+                        return oldSolution.WithDocumentText(documentId, newText, PreservationMode.PreserveValue);
+                    }
+                },
+                data: (@this: this, documentId, textContainer, isCurrentContext),
+                onAfterUpdate: static (oldSolution, newSolution, data) =>
                 {
-                    // We don't have the old text or version.  Rather than trying to reuse a version that we still have, let's just assume the file has changed.
-                    // keep open document text alive by using PreserveIdentity
-                    currentSolution = oldSolution.WithDocumentText(documentId, newText, PreservationMode.PreserveValue);
-                }
+                    var (@this, documentId, textContainer, isCurrentContext) = data;
 
-                var newSolution = this.SetCurrentSolution(currentSolution);
-                SignupForTextChanges(documentId, textContainer, isCurrentContext, (w, id, text, mode) => w.OnDocumentTextChanged(id, text, mode));
+                    @this.AddToOpenDocumentMap(documentId);
+                    @this.SignupForTextChanges(documentId, textContainer, isCurrentContext, (w, id, text, mode) => w.OnDocumentTextChanged(id, text, mode));
 
-                var newDoc = newSolution.GetRequiredDocument(documentId);
-                this.OnDocumentTextChanged(newDoc);
+                    var newDoc = newSolution.GetRequiredDocument(documentId);
+                    @this.OnDocumentTextChanged(newDoc);
 
-                // Fire and forget that the workspace is changing.
-                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
+                    // Fire and forget that the workspace is changing.
+                    @this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
 
-                // We fire 2 events on source document opened.
-                this.RaiseDocumentOpenedEventAsync(newDoc);
-                this.RaiseTextDocumentOpenedEventAsync(newDoc);
-            }
+                    // We fire 2 events on source document opened.
+                    @this.RaiseDocumentOpenedEventAsync(newDoc);
+                    @this.RaiseTextDocumentOpenedEventAsync(newDoc);
+                });
 
+            // TODO: why is this here, and not in onAfterUpdate?
```

Comment:
....that's a good question. :-/

Reply:
maybe it doesn't want to run undre a lock?  but i can't see any issues that would arise if we moved this under the locked section.